ts: Time series memory monitoring

mrtracy opened the pull request:
Outfit the time series query system with memory monitoring.

+ TS Server creates an unlimited memory monitor. Incoming requests
create monitors using the top-level monitor as a pool.
+ TS server passes a BoundAccount to each invocation of DB.Query.
+ DB.Query method applies its major sources of memory growth to
BoundAccount.Grow().

The TS Server monitor is currently unlimited, but a limit will be
applied once resource governing is in place.

Release note: None

Commit:
ts: Time series memory monitoring

Outfit the time series query system with memory monitoring.

+ TS Server creates an unlimited memory monitor. Incoming requests
create monitors using the top-level monitor as a pool.
+ TS server passes a BoundAccount to each invocation of DB.Query.
+ DB.Query method applies its major sources of memory growth to
BoundAccount.Grow().

The TS Server monitor is currently unlimited, but a limit will be
applied once resource governing is in place.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21637)
<!-- Reviewable:end -->

mrtracy commented:
@knz, can you let me know if i'm on the right track here?  I'm going to be adding some memory-related governance to TS queries for #20014, but I'm adding this system in as well to more accurately track memory (and abort queries if the governance system fails).

knz commented:
Good to see this happening. As these things go, it would be nice to see a test that proves the code errors out if a monitor limit is reached.

Also I think there's something to discuss conceptually, see below.
Cheers

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query.go, line 913 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SV7ZC4GJqjhydr1fA:-L3SV7ZC4GJqjhydr1fB:b-udmv18) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L913)):*
> ```Go
> 	}
> 
> 	if err := sessionAccount.Grow(ctx, int64(sizeOfDataPoint*len(responseData))); err != nil {
> ```

Here and above -- conceptually you're missing a point. The call to `Grow` should be performed *before* you construct the data that occupies memory. If you hit a limit, the monitor must tell you *before* you hit the heap allocator and possibly kill your process with a OOM error.

---

*[pkg/ts/query.go, line 939 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SVPOhcJ2xvkVLMB2X:-L3SVPOhcJ2xvkVLMB2Y:b-k246na) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L939)):*
> ```Go
> 		}
> 		if _, ok := sourceSpans[source]; !ok {
> 			sourceSpans[source] = &dataSpan{
> ```

Here `Grow` should be called. Pass the account as argument to this function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3SVaQpm82zhjC7V5eu:b-dqq6zz)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'm mildly apprehensive about the memory monitoring as it seems so easy to not account for something and there is no easy way to test that we are capturing all the memory usage. I wonder if we can have a test that does a few simple queries in a loop and measures the memory usage. The memory usage can be retrieved using `runtime.ReadMemStats` and looking at the `MemStats.TotalAllocs` field (this is what `go test -bench` uses). 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query.go, line 861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3TIo_V_uNF-15TcQWx:-L3TIo_V_uNF-15TcQWy:bihz86p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L861)):*
> ```Go
> 	for name, span := range sourceSpans {
> 		sources = append(sources, name)
> 		totalSizeOfSources += sizeOfRune * len(name)
> ```

`name` is UTF-8 encoded so `len(name)` is the number of bytes it consumes. There is no need to multiply by `sizeOfRune` here.  You have to cast to `[]rune(name)` if you want a slice of runes (which you don't, I'm just pointing out the difference between runes and UTF-8).

---

*[pkg/ts/server.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3THp8KRKsfO5XcO4K9:-L3THp8KRKsfO5XcO4KA:b1zpaus) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/server.go#L43)):*
> ```Go
> 	// time series queries. This is not currently enforced, but is used for
> 	// monitoring purposes.
> 	queryMemoryMax = 500 * 1024 * 1024 // 500MiB
> ```

Just skimming this change and saw this bit. This value seems super-high. How much memory do you expect a typical 30-day timeseries query to use? I would have naively expected this to be 64MiB or less.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3TJBQzE0SFnLVzAIrA:br6715k)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
The issue isn't memory usage, it's maximum memory usage at a single time; even with these adjustments, the number of total allocations will be the same.  However, currently the system needs the entire hunk of data referenced in memory at the same time; after the adjustment, it will only need a portion of the data in memory at any given time.  It still has to process the same amount of data.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query.go, line 861 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3TIo_V_uNF-15TcQWx:-L3U-LptsnaPCt7WaZvy:b-cwgaxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L861)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`name` is UTF-8 encoded so `len(name)` is the number of bytes it consumes. There is no need to multiply by `sizeOfRune` here.  You have to cast to `[]rune(name)` if you want a slice of runes (which you don't, I'm just pointing out the difference between runes and UTF-8).
</blockquote></details>

Thanks, I wasn't clear on that.

---

*[pkg/ts/query.go, line 913 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SV7ZC4GJqjhydr1fA:-L3U-NYhkCvJ8g6Gfn0U:b-yjg8vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L913)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here and above -- conceptually you're missing a point. The call to `Grow` should be performed *before* you construct the data that occupies memory. If you hit a limit, the monitor must tell you *before* you hit the heap allocator and possibly kill your process with a OOM error.
</blockquote></details>

The issue is that I don't know how much memory is being used until it's already being used; all of the data comes from the KV layer, and i'm just counting it here.

My plan is to have resource governance *before* using the memory management, using the following strategy:

1. Compute a "soft-limit" of memory for use by the time series query system, and reserve it.
2. We have a limit on the number of simultaneous workers servicing queries (currently 250, might be adjusted); I will divide the soft limit evenly by this number  to determine the budget per-worker.
3. Queries for large time spans may be broken up among multiple workers if the amount of memory for a single query would exceed the per-worker budget.
4. Each worker will reserve some space from the soft-limit by *estimating* how much memory it will need.  This estimation will include the total number of active nodes on the cluster.
5. The workers will use still use the "Grow" mechanism to catch places where the estimation is somehow wrong and abort queries if necessary.

Essentially, the system is designed such that:
1. Ideally, no query should ever be aborted due to memory constraints due to the first-level resource governance
2. If thats resource governance fails, queries will be aborted by the memory management system, which will be noticeable in our logs.

---

*[pkg/ts/query.go, line 939 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SVPOhcJ2xvkVLMB2X:-L3U0kh5tARIaJdTwDNE:b-my286f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L939)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here `Grow` should be called. Pass the account as argument to this function.
</blockquote></details>

Is it okay from a performance perspective to call 'Grow' so many times? I thought it better to count and call Grow one time, since the majority of this memory is already being used.

---

*[pkg/ts/server.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3THp8KRKsfO5XcO4K9:-L3U0wGL5sJI7D5c4kbK:b-t72nk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Just skimming this change and saw this bit. This value seems super-high. How much memory do you expect a typical 30-day timeseries query to use? I would have naively expected this to be 64MiB or less.
</blockquote></details>

This is the limit for the entire time series query.  A single 30 day query will use about 10MB *per node*.  However, one page of our Admin UI might launch 30-40 queries at a single time; on a six node cluster, this is were you start to see gigabytes of memory being used at once.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3U1AxmAsmxkPWMNVq6:b-po89qe)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query.go, line 913 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SV7ZC4GJqjhydr1fA:-L3UDR-sFpe2-ONGRpMc:bknkiuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L913)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

The issue is that I don't know how much memory is being used until it's already being used; all of the data comes from the KV layer, and i'm just counting it here.

My plan is to have resource governance *before* using the memory management, using the following strategy:

1. Compute a "soft-limit" of memory for use by the time series query system, and reserve it.
2. We have a limit on the number of simultaneous workers servicing queries (currently 250, might be adjusted); I will divide the soft limit evenly by this number  to determine the budget per-worker.
3. Queries for large time spans may be broken up among multiple workers if the amount of memory for a single query would exceed the per-worker budget.
4. Each worker will reserve some space from the soft-limit by *estimating* how much memory it will need.  This estimation will include the total number of active nodes on the cluster.
5. The workers will use still use the "Grow" mechanism to catch places where the estimation is somehow wrong and abort queries if necessary.

Essentially, the system is designed such that:
1. Ideally, no query should ever be aborted due to memory constraints due to the first-level resource governance
2. If thats resource governance fails, queries will be aborted by the memory management system, which will be noticeable in our logs.
</blockquote></details>

yes that sounds good. Worth spelling out in the commit message for future reference though. Perhaps as a comment around here too.

---

*[pkg/ts/query.go, line 939 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SVPOhcJ2xvkVLMB2X:-L3UDWuOJmUy88LX7HNU:b-cu1w67) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L939)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Is it okay from a performance perspective to call 'Grow' so many times? I thought it better to count and call Grow one time, since the majority of this memory is already being used.
</blockquote></details>

Yes `Grow` is meant to be used with fine granularity. The "costly" operation to go one up in the memory monitor is only done once every "block size" allocations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> The issue isn't memory usage, it's maximum memory usage at a single time; even with these adjustments, the number of total allocations will be the same. However, currently the system needs the entire hunk of data referenced in memory at the same time; after the adjustment, it will only need a portion of the data in memory at any given time. It still has to process the same amount of data.

Fair point that my test suggestion won't work as presented. My point still stands that the memory accounting in this manner is fragile so we should take some steps to ensure that we're testing that the significant memory allocations are captured.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ts/server.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3THp8KRKsfO5XcO4K9:-L3UHeV4y2Dio9R4Cdcx:b-roua1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This is the limit for the entire time series query.  A single 30 day query will use about 10MB *per node*.  However, one page of our Admin UI might launch 30-40 queries at a single time; on a six node cluster, this is were you start to see gigabytes of memory being used at once.
</blockquote></details>

Right. We will need limit concurrency in timeseries query processing to keep under some max memory usage. Using up to 500MiB seems excessive.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3UGyzJxSiLoSguYzGn:bcfhopq)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Yes, I think I misunderstood you.  It's difficult to verify maximum memory usage explicitly in a repeatable test, but I'll see if I can utilize MemStats in some sort of test to verify that we are accounting for everything - this would be more of a sanity check than an exact number, though.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ts/server.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3THp8KRKsfO5XcO4K9:-L3UK3N1UHlcSjiK_v_y:bfh7wy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Right. We will need limit concurrency in timeseries query processing to keep under some max memory usage. Using up to 500MiB seems excessive.
</blockquote></details>

Are you okay with a smaller constant, or would you prefer to derive a number from max physical memory?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3UJ2s17y0OVcmZu6c1:b-s7yrm1)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ts/server.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3THp8KRKsfO5XcO4K9:-L3UKlVi5i-8lxhN2SXi:b3iwikl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Are you okay with a smaller constant, or would you prefer to derive a number from max physical memory?
</blockquote></details>

I want to see a smaller constant. I'd be comfortable with something on the order of 64MiB. As it is, this constant is larger than the default cache size we give to RocksDB and the max-sql memory setting.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
I've adjusted the soft limit, and taken Raphael's suggestion of making calls to `Grow` more granular.  I've also added a test to verify that memory pressure causes queries to abort.

In terms of verifying that the memory allocations match reality, I think we should use the same strategy used to verify sql memory management: one of the command-line TCL tests. It is an acceptance test that sets a very low memory limit for a running cockroachdb instance, and then executes a memory-intensive operation that would crash the process if memory  management were not running.

I can write a small stand-alone command line application to run a memory-intensive time series operation against cockroach. This would allow me to use the roughly the same test outline to verify memory management for the time series system.

That said, I don't think this PR should be blocked on creating this test.

---

Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query.go, line 913 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SV7ZC4GJqjhydr1fA:-L3_1ttM8sQevu4QpV0D:b-278e3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L913)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

yes that sounds good. Worth spelling out in the commit message for future reference though. Perhaps as a comment around here too.
</blockquote></details>

I've added this comment to issue #20014, which is driving this PR and others.

---

*[pkg/ts/query.go, line 939 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3SVPOhcJ2xvkVLMB2X:-L3_0pxLlgSit5NwNHTt:bfbw5tj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/query.go#L939)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes `Grow` is meant to be used with fine granularity. The "costly" operation to go one up in the memory monitor is only done once every "block size" allocations.
</blockquote></details>

Okay, i am now calling Grow() in a granular fashion everywhere, no more totals are maintained.

---

*[pkg/ts/server.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3THp8KRKsfO5XcO4K9:-L3_0wbkDksew5HUz2h6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2495ab6259d1f7b88aee69efef72deb6c8a53016/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I want to see a smaller constant. I'd be comfortable with something on the order of 64MiB. As it is, this constant is larger than the default cache size we give to RocksDB and the max-sql memory setting.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3_2-BNHY2jMiAjYNyX:bmc3wfq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> In terms of verifying that the memory allocations match reality, I think we should use the same strategy used to verify sql memory management: one of the command-line TCL tests. It is an acceptance test that sets a very low memory limit for a running cockroachdb instance, and then executes a memory-intensive operation that would crash the process if memory management were not running.

Generating the data for this test will complicate writing it. I think it will be easier to do something that is a normal Go test and use `runtime.ReadMemStats` to check allocations. I agree this shouldn't block this PR. We're getting caught up in this when the priority is to get the excessive memory usage fixed, not monitor it. 

---

Review status: 1 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query_test.go, line 1193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3c9D1OeJPrikTSOd7E:-L3c9D1P07rXIf_1Xkh6:b6fd3wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/299161d461b673dc1c988dea49713f19a6d419a0/pkg/ts/query_test.go#L1193)):*
> ```Go
> 
> 	queryAcc := limitedMon.MakeBoundAccount()
> 	defer queryAcc.Close(context.TODO())
> ```

Even though you can't assert anything, I think you should make a call to `runtime.ReadMemStats` here and immediately after the `DB.Query` call and output how much memory was used. 

---

*[pkg/ts/query_test.go, line 1197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3c9tO2XWCT0p5-oEQt:-L3c9tO2XWCT0p5-oEQu:bib6re1) ([raw file](https://github.com/cockroachdb/cockroach/blob/299161d461b673dc1c988dea49713f19a6d419a0/pkg/ts/query_test.go#L1197)):*
> ```Go
> 		context.TODO(), tspb.Query{Name: "test.metric"}, resolution1ns, 1, 0, 10000, 0, &queryAcc,
> 	)
> 	if err == nil {
> ```

Is this necessary given the `testutils.IsError` check below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-:-L3cA8rTMfVsFeBXInsR:b-q0d5im)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 1 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ts/query_test.go, line 1193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3c9D1OeJPrikTSOd7E:-L3dX2Z-VV5mPVnQ68nP:b-x2caqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/299161d461b673dc1c988dea49713f19a6d419a0/pkg/ts/query_test.go#L1193)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Even though you can't assert anything, I think you should make a call to `runtime.ReadMemStats` here and immediately after the `DB.Query` call and output how much memory was used. 
</blockquote></details>

Alright, i've installed a section at the end of this test that uses the ReadMemStats strategy to compare the allocations before and after the query.  I've also compared this with the Maximum allocation level noted by the BytesMonitor.  The current output looks like this:

```
total allocations for query: 17944
maximum allocations for query monitor: 800
```

As a one-time thing, I also added a similar ReadMemStats block directly into the Query method, in order to measure the allocations in the KV layer that are not tracked by the memory monitor.  That looks like this:

```
!!!QUERY TOTALALLOCS FOR KV LAYER: 14824
total allocations for query: 17944
maximum allocations for query monitor: 800
```

As you can see, the KV layer does account for most of the bytes being allocated, as is expected.  Not all of these allocated structures are being returned to the Query method and thus are not measured.

We can at least monitor these numbers to make sure they don't increase unexpectedly later.

---

*[pkg/ts/query_test.go, line 1197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3c9tO2XWCT0p5-oEQt:-L3dPSlEAHp0ZX8KeySe:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/299161d461b673dc1c988dea49713f19a6d419a0/pkg/ts/query_test.go#L1197)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this necessary given the `testutils.IsError` check below?
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ts/query_test.go, line 1193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637#-L3c9D1OeJPrikTSOd7E:-L3d_7DvAf1IQMbUwVTS:brd4a8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/299161d461b673dc1c988dea49713f19a6d419a0/pkg/ts/query_test.go#L1193)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Alright, i've installed a section at the end of this test that uses the ReadMemStats strategy to compare the allocations before and after the query.  I've also compared this with the Maximum allocation level noted by the BytesMonitor.  The current output looks like this:

```
total allocations for query: 17944
maximum allocations for query monitor: 800
```

As a one-time thing, I also added a similar ReadMemStats block directly into the Query method, in order to measure the allocations in the KV layer that are not tracked by the memory monitor.  That looks like this:

```
!!!QUERY TOTALALLOCS FOR KV LAYER: 14824
total allocations for query: 17944
maximum allocations for query monitor: 800
```

As you can see, the KV layer does account for most of the bytes being allocated, as is expected.  Not all of these allocated structures are being returned to the Query method and thus are not measured.

We can at least monitor these numbers to make sure they don't increase unexpectedly later.
</blockquote></details>

Thanks. The KV layer is a pig. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21637)*
<!-- Sent from Reviewable.io -->
